### PR TITLE
RavenDB-17599: Implementation of Highlights

### PR DESCRIPTION
**Pull request by redknightlois**
I would need some help here, it seems that I am hitting some unfinished implementation issues.

### CODE REVIEW

**Comment by arekpalinski** [comment]
@redknightlois FYI the PR with static indexes got merged https://github.com/ravendb/ravendb/pull/13730. So you can check the highlighting with their usage.

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 1]
```diff
@@ -184,13 +192,16 @@ private void SetupHighlighter(IndexQueryServerSide query, JsonOperationContext c
...
 
                 if (_tagsPerField == null)
-                    _tagsPerField = new Dictionary<string, (string[] PreTags, string[] PostTags)>();
+                    _tagsPerField = new();
 
                 var fieldName = query.Metadata.IsDynamic
                     ? throw new NotSupportedException("AutoIndex dynamic field is not supported yet.")
                     : highlighting.Field.Value;
+                
+                throw new NotImplementedException();
```
Comment (redknightlois) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Can anyone give me a hint on how on earth can I get the actual terms I am looking for to highlight later? I understand the complexity of the `Query` but it seems to me that we are missing a few levels of separation from Lucene (or I am not understanding the abstraction). 

cc @maciejaszyk

Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Let's talk about that directly.

Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: https://github.com/maciejaszyk/ravendb/tree/RavenDB-17599.terms @redknightlois

**Comment by redknightlois** [comment]
@maciejaszyk The rebase from the old code was a mess, after trying multiple times to rebase gave up and reconstruct it from your commit instead. Can you take a look I am not missing anything?

**Comment by maciejaszyk** [comment]
@redknightlois I rebased this PR to the new evaluator.  Also did some work at `CoraxIndexReadOperation` and fetched terms for you.
image
https://github.com/maciejaszyk/ravendb/tree/RavenDB-17599.QueryBuilder

**Comment by arekpalinski** [comment]
Can you please do the rebase so the conflicts should disappear?

**Comment by arekpalinski** [comment]
In general, what the status here?

**Comment by redknightlois** [comment]
> In general, what the status here?

We had a long call with @maciejaszyk about some edge cases I found. We believe we found the proper way to address them, given how the code surrounding is changing given the latest work been done.

**Comment by arekpalinski** [comment]
@redknightlois please check the NRE thrown in tests

**Comment by redknightlois** [comment]
We are still missing some work on it, that's why it is still in draft. It appears that we have still some missing wire-up up in the chain that @maciejaszyk is gonna look into (most likely related to map-reduce queries).

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 2]
```diff
@@ -171,6 +289,148 @@ QueryResult GetQueryResult(Document document, ref bool markedAsSkipped)
...
+                    // We will look for a whitespace after the match to end the token. 
+                    int tokenEnd = fieldFragment.Slice(index)
+                                                .IndexOf(' ');
+                    if (tokenEnd < 0)
+                        tokenEnd = fieldFragment.Length - index;
+
+                    tokenEnd += index; // Adjusting to absolute positioning
+
+                    int highlightingLength = preTag.Length + postTag.Length + (tokenEnd - tokenStart);
+
```
Comment (redknightlois) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: There is an alternative interpretation here for the length. Should we include the size of the pre and post tag as part of the fragment size? cc @arekpalinski @ayende

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: I don't think so.
It doesn't actually matter usually.

**Comment by arekpalinski** [comment]
@redknightlois Can you please check the conflicts?

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryEvaluator.cs [hunk 3]
```diff
@@ -0,0 +1,1277 @@
+﻿using System;
```
Comment (maciejaszyk) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryEvaluator.cs]: This is leftover. Currently, we're using `CoraxQueryBuilder`.

File: test/SlowTests/Issues/RavenDB_11906.cs [hunk 4]
```diff
@@ -129,7 +129,6 @@ public void SupportForCreateFieldWithOptions(Options options)
 
         [Theory]
         [RavenData(SearchEngineMode = RavenSearchEngineMode.Lucene)]
-        [RavenData(SearchEngineMode = RavenSearchEngineMode.Corax, Skip = "RavenDB-17966")]
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB_11906.cs:132]: Shouldn't we also set `SearchEngineMode = RavenSearchEngineMode.All` now?

File: test/SlowTests/Issues/RavenDB_6558.cs [hunk 5]
```diff
@@ -14,19 +15,19 @@ public RavenDB_6558(ITestOutputHelper output) : base(output)
         {
         }
 
-        [Fact]
-        public void CanUseDifferentPreAndPostTagsPerField()
+        [RavenTheory(RavenTestCategory.Highlighting)]
+        [RavenData(SearchEngineMode = RavenSearchEngineMode.Corax)]
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB_6558.cs]: Only Corax?

File: test/SlowTests/Tests/Querying/HighlightesTests.cs [hunk 6]
```diff
@@ -66,10 +67,10 @@ public ContentSearchIndex()
         }
 
         [Theory]
-        [InlineData("session")]
-        public void SearchWithHighlightes(string q)
+        [RavenData("session", SearchEngineMode = RavenSearchEngineMode.Lucene)]
```
Comment (arekpalinski) [inline test/SlowTests/Tests/Querying/HighlightesTests.cs]: Only Lucene?

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 7]
```diff
@@ -150,19 +176,111 @@ QueryResult GetQueryResult(Document document, ref bool markedAsSkipped)
...
+                                        int groupKey;
+                                        if ((groupKey = document.Data.GetPropertyIndex(fieldDescription.GroupKey)) != -1)
+                                        {
+                                            document.Data.GetPropertyByIndex(groupKey, ref property);
+
+                                            key = property.Token switch
+                                            {
+                                                BlittableJsonToken.String => ((LazyStringValue)property.Value).ToString(),
+                                                BlittableJsonToken.CompressedString => ((LazyCompressedStringValue)property.Value).ToString(),
+                                                _ => throw new Exception("If this happens, we have a bug.")
```
Comment (maciejaszyk) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Please change the exception message.

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 8]
```diff
@@ -150,19 +176,111 @@ QueryResult GetQueryResult(Document document, ref bool markedAsSkipped)
...
+                                            key = document.Id;
+                                        }
+                                    }
+                                    else
+                                        key = document.Id;
+
+                                    
+                                    if (tokensDictionary.TryGetValue(key, out var existingHighlights))
+                                    {
+                                        throw new Exception("??? Please look at LuceneIndexReadOperation");
```
Comment (maciejaszyk) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Like above. I've made it for testing only.

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs [hunk 9]
```diff
@@ -171,6 +289,149 @@ QueryResult GetQueryResult(Document document, ref bool markedAsSkipped)
...
+                        for (int i = 0; i < ls.Count; i++)
+                            nls[i] = ls[i].TrimEnd('*').TrimStart('*');
+                        break;
+                    case Tuple<string, string> t2:
+                        nls = new string[] { t2.Item1.TrimEnd('*').TrimStart('*'), t2.Item2.TrimEnd('*').TrimStart('*') };
+                        break;
+                    case string[] as1:
+                        continue;
+                    default:
+                        throw new Exception("If this happens, we have a bug.");
```
Comment (maciejaszyk) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs]: Like above.

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs [hunk 10]
```diff
@@ -88,6 +88,8 @@ public override long EntriesCount()
         {
             ExplanationOptions explanationOptions = null;
 
+            // The reason why we keep distincts counts here, is because at the Lucene level there is no sorting unique accesse like we have
```
Comment (maciejaszyk) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs:91]: Should we keep them? @arekpalinski

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 11]
```diff
@@ -1131,7 +1131,7 @@ internal static int GetEntry(CompactTree tree, Page page, ushort entriesOffset,
             encodedKey.Key.CopyTo(output.ToSpan());
 
             var outputSpan = output.ToSpan();
-            key = output[^1] == 0 ? outputSpan : outputSpan.Slice(0, outputSpan.Length - 1);
+            key = outputSpan[^1] == 0 ? outputSpan[0..^1] : outputSpan;
```
Comment (maciejaszyk) [inline src/Voron/Data/CompactTrees/CompactTree.cs:1217]: AFAIR we sliced the buffer to delete the null character (`\0`).

File: src/Voron/Slice.Extensions.cs [hunk 12]
```diff
@@ -18,11 +18,7 @@ public static bool StartWith(this Slice s1, ReadOnlySpan<byte> s2)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]        
         public static bool EndsWith(this Slice s1, ReadOnlySpan<byte> s2)
         {
-            var term = s1.AsReadOnlySpan();
-            
-            return term.Length > 1 
-                ? s1.AsReadOnlySpan()[..(term.Length - 1)].EndsWith(s2)
-                : s1.AsReadOnlySpan().EndsWith(s2);
+            return s1.AsReadOnlySpan().EndsWith(s2);
```
Comment (maciejaszyk) [inline src/Voron/Slice.Extensions.cs:21]: Like above. Did you handle that case?

Comment (maciejaszyk) [reply src/Voron/Slice.Extensions.cs:21]: For example when you have indexed `term`.
So, s1 is written as `term\0` but s2 is `term`.

File: test/SlowTests/Issues/RavenDB_11906.cs [hunk 13]
```diff
@@ -129,7 +129,6 @@ public void SupportForCreateFieldWithOptions(Options options)
 
         [Theory]
         [RavenData(SearchEngineMode = RavenSearchEngineMode.Lucene)]
```
Comment (maciejaszyk) [inline test/SlowTests/Issues/RavenDB_11906.cs:131]: `[RavenData(SearchEngineMode = RavenSearchEngineMode.All)]`

File: test/SlowTests/Issues/RavenDB-10343.cs [hunk 14]
```diff
@@ -552,7 +552,7 @@ public async Task CanQueryWithLoadFromSelectAndProjectWhereAsync(Options options
         }
 
         [Theory]
-        [RavenData(SearchEngineMode = RavenSearchEngineMode.All)]
+        [RavenData(SearchEngineMode = RavenSearchEngineMode.Corax)]
```
Comment (maciejaszyk) [inline test/SlowTests/Issues/RavenDB-10343.cs]: `[RavenData(SearchEngineMode = RavenSearchEngineMode.All)]`

File: test/SlowTests/Issues/RavenDB_6558.cs (hunk 5 above)
Comment (maciejaszyk) [inline test/SlowTests/Issues/RavenDB_6558.cs]: `[RavenData(SearchEngineMode = RavenSearchEngineMode.All)]`

File: test/SlowTests/Tests/Querying/HighlightesTests.cs (hunk 6 above)
Comment (maciejaszyk) [inline test/SlowTests/Tests/Querying/HighlightesTests.cs]: `[RavenData(SearchEngineMode = RavenSearchEngineMode.All)]`

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs (hunk 10 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs:91]: IMO it's okay to leave it

File: test/SlowTests/Issues/RavenDB_11906.cs (hunk 13 above)
Comment (redknightlois) [reply test/SlowTests/Issues/RavenDB_11906.cs:131]: No, dynamic fields are not supported yet.

File: test/SlowTests/Issues/RavenDB_11906.cs (hunk 4 above)
Comment (redknightlois) [reply test/SlowTests/Issues/RavenDB_11906.cs:132]: No, dynamic fields are not supported yet.

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 11 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs:1217]: That is been done here to be consistent on usage.

File: src/Voron/Slice.Extensions.cs (hunk 12 above)
Comment (redknightlois) [reply src/Voron/Slice.Extensions.cs:21]: There is no reason why you should be doing that here... I dont know what I was thinking when I wrote this code. Compact Tree development era.

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs (hunk 10 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs:91]: This is an explanation on what are the differences. This type of information it is invaluable when you need to debug some issue.

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryEvaluator.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryEvaluator.cs]: @redknightlois can you please delete that file?

File: test/SlowTests/Issues/RavenDB_11906.cs (hunk 4 above)
Comment (arekpalinski) [reply test/SlowTests/Issues/RavenDB_11906.cs:132]: Okay. So we can update the issue number here - dynamic fields is "RavenDB-18144"

**Comment by arekpalinski** [comment]
@redknightlois Can you please resolve the conflicts?

**Comment by redknightlois** [comment]
Done

**Comment by redknightlois** [comment]
I dont understand why it sais that it has conflicts... If it comes directly from the current `/features/corax` commit.

**Comment by arekpalinski** [comment]
@redknightlois I did the rebase and indeed I got multiple conflicts. I tried to solve them and opened https://github.com/ravendb/ravendb/pull/14179 to verify if I did it correctly.

**Comment by redknightlois** [comment]
Found the problem... 🤦 
I had the authoritative hidden, so it was my local branch not the actual branch. Rebasing was not so easy, use this one instead.

**Pull request merged**